src/main/plugin/iso19139.ca.HNAP/layout/tpl-csv.xsl

```diff
@@ -0,0 +1,210 @@
+        </xsl:choose>
+      </category>
+      <metadatacreationdate>
+        <xsl:value-of select="gmd:dateStamp/*"/>
+      </metadatacreationdate>
+      <referencesystem>
+        <xsl:value-of select="gmd:referenceSystemInfo/gmd:MD_ReferenceSystem/gmd:referenceSystemIdentifier/gmd:RS_Identifier/gmd:code/gco:CharacterString"/>
+      </referencesystem>
+
+      <!-- For each date grouped by the date type mapped to a readable value -->
+      <xsl:for-each select="gmd:identificationInfo/*/gmd:citation/*/gmd:date">
+        <xsl:variable name="dateTypeCode" select="*/gmd:dateType/*/@codeListValue"/>
+        <xsl:variable name="dateTypeCodeReadable" select="tokenize($codelists/codelist[@name = 'gmd:CI_DateTypeCode']/entry[code/text() = $dateTypeCode]/value/text(), ';')[1]"/>
+
+        <xsl:element name="date-{$dateTypeCodeReadable}">
+          <xsl:value-of select="*/gmd:date/*/text()"/>
+        </xsl:element>
+      </xsl:for-each>
+
+      <xsl:for-each select="gmd:identificationInfo/*/gmd:graphicOverview/*/gmd:fileName">
+        <image>
+          <xsl:value-of select="*/text()"/>
+        </image>
+      </xsl:for-each>
+
+
+      <!-- All keywords not having thesaurus reference or an empty thesaurusName -->
+      <xsl:for-each select="gmd:identificationInfo/*/gmd:descriptiveKeywords/gmd:MD_Keywords[not(gmd:thesaurusName) or gmd:thesaurusName/*/gmd:title/(gco:CharacterString|gmx:Anchor)/text() = '']">
+        <xsl:variable name="keywordTypeCode" select="gmd:type/*/@codeListValue"/>
+        <xsl:variable name="keywordTypeCodeReadable" select="tokenize($codelists/codelist[@name = 'gmd:MD_KeywordTypeCode']/entry[code/text() = $keywordTypeCode]/value/text(), ';')[1]"/>
+
+        <xsl:for-each select="gmd:keyword[not(@gco:nilReason) or */text() != '']">
+          <xsl:choose>
+            <!-- All keywords with a valid keywordTypeCode -->
+            <xsl:when test="$keywordTypeCodeReadable != ''">
+              <xsl:element name="keyword-{$keywordTypeCodeReadable}">
+                <xsl:apply-templates mode="localised" select=".">
+                  <xsl:with-param name="langId" select="$langId"/>
+                </xsl:apply-templates>
+              </xsl:element>
+            </xsl:when>
+
+            <!-- All keywords without a valid keywordTypeCode -->
+            <xsl:otherwise>
+              <keyword-other>
+                <xsl:apply-templates mode="localised" select=".">
+                  <xsl:with-param name="langId" select="$langId"/>
+                </xsl:apply-templates>
+              </keyword-other>
+            </xsl:otherwise>
+          </xsl:choose>
+        </xsl:for-each>
+      </xsl:for-each>
+
+      <!-- All keywords with a valid thesaurus name -->
+      <xsl:for-each select="gmd:identificationInfo/*/gmd:descriptiveKeywords/gmd:MD_Keywords[gmd:thesaurusName/*/gmd:title/(gco:CharacterString|gmx:Anchor)/text() != '']">
+
+        <!-- Get the localeId matching the english LanguageCode (eng) -->
```

Comment:
Maybe I'm not understanding what you are trying to do?  Generally you should not need English - generally you need the main language and alternative language

I'm not sure why you are not using `$langId` like all the other calls to `mode="localised"`?

`<xsl:with-param name="langId" select="$langId"/>`

If you need to get the main language or alternative language, then you may want to look at the following

https://github.com/metadata101/iso19139.ca.HNAP/blob/085ee01500873bb768dfef69e0a36814a941944b/src/main/plugin/iso19139.ca.HNAP/index-fields/default.xsl#L38-L63

It gets a main language and an alternate language from the metadata record.

Reply:
What I'm trying to do is make the CSV export column headers more readable. And in this context specifically making the thesaurus names in the keyword column headers more descriptive.

So instead of `keyword-geonetworkthesaurusexternalthemeGCCoreSubject` ([already always in English](https://github.com/metadata101/iso19139.ca.HNAP/pull/374#discussion_r1549743144)) derived from the thesaurus identifier code I wanted to use the English translation of the thesaurus title (`keyword-GovernmentofCanadaCoreSubjectThesaurus`).

As all the other column headers are always in English regardless of metadata main language or UI language ([see testing results above](https://github.com/metadata101/iso19139.ca.HNAP/pull/374#discussion_r1549743144)) I thought it made sense to force the use of the English translation of the title. The only thing that is translated right now in prod is the content for each column but this has errors as well when the document uses fra instead of fre (fixed in my code).

I could use `$langId` to translate the thesaurus title into the UI language but then the column headers will be inconsistent as all the other headers are English. I don't think the main language or alt language of the documents would help with this.

Reply:
Ok that makes sense, I didn't have enough sample data to catch those cases. Maybe we should continue to use the identifier code for the thesaurus then. Would it be safe to strip out the `geonetworkthesaurusexternaltheme` so `keyword-geonetworkthesaurusexternalthemeGCCoreSubject` -> `keyword-GCCoreSubject`

The French character issues seem to just be with excel not knowing the encoding standard. If imported into excel with UTF-8 it seems fine. Opening the raw text also looks fine.

I did notice another bug where double-quotes are escaped using backslashes instead of another double quote resulting in excel and other software parsing the columns incorrectly. However this would be a change for geonetwork-core. Is there a reason backslashes are used instead of quotes?

[RFC 4180](https://www.ietf.org/rfc/rfc4180.txt#:~:text=If%20double-quotes%20are%20used%20to%20enclose%20fields%2c%20then%20a%20double-quote%0a%20%20%20%20%20%20%20appearing%20inside%20a%20field%20must%20be%20escaped%20by%20preceding%20it%20with%0a%20%20%20%20%20%20%20another%20double%20quote.): If double-quotes are used to enclose fields, then a double-quote appearing inside a field must be escaped by preceding it with another double quote.

Reply:
Correct, the identifier code is more reliable.

Looking at sample data, it looks something like this.

```
<gmd:identifier>
<gmd:MD_Identifier>
<gmd:code>
<gmx:Anchor xlink:href="http://localhost/geonetwork/srv/api/registries/vocabularies/external.theme.GC_Core_Subject">geonetwork.thesaurus.external.theme.GC_Core_Subject</gmx:Anchor>
</gmd:code>
</gmd:MD_Identifier>
</gmd:identifier>
```

Note that it is possible to upload a thesaurus in themes called myThesaurus and then we can upload another thesaurus in place called myThesaurus.
So they would be external.theme.myThesaurus and external.place.myThesaurus

So the categories(theme/place) need to remain to be unique.

Also same applies for the type of thesaurus as you can see documented here there are different types that can be applied.
https://docs.geonetwork-opensource.org/3.12/administrator-guide/managing-classification-systems/managing-thesaurus/

Regarding geonetwork.thesaurus, keep in mind that geonetwork is just a tool that manages the thesaurus. There is nothing stopping someone from hosting their own thesaurus and using something like 

```
<gmd:identifier>
<gmd:MD_Identifier>
<gmd:code>
<gmx:Anchor xlink:href="http://myhost.com/mythesaurus/external.theme.GC_Core_Subject">external.theme.GC_Core_Subject</gmx:Anchor>
</gmd:code>
</gmd:MD_Identifier>
</gmd:identifier>
```

In this case, if  `geonetwork.thesaurus` is removed then this would conflict with this entry - however, I'm not sure why someone would host their own thesaurus and call it the same name as one hosted on the geonetwork but I guess it could happen.

So keeping the full name is more reliable.  

If anything, there could be a bug with `geonetwork.thesaurus` because if we have multiple geonetwork instances deployed and we create a record where one geonetwork instances points to the thesaurus of another geonetwork instance then it is possible that we have the following 2 entries.

<gmx:Anchor xlink:href="http://localhost/geonetwork1/srv/api/registries/vocabularies/external.theme.GC_Core_Subject">geonetwork.thesaurus.external.theme.GC_Core_Subject</gmx:Anchor>

<gmx:Anchor xlink:href="http://localhost/geonetwork2/srv/api/registries/vocabularies/external.theme.GC_Core_Subject">geonetwork.thesaurus.external.theme.GC_Core_Subject</gmx:Anchor>

However since the record from geonetwork2 would have been added manually, the users who added the entry should have called it `geonetwork2.thesaurus.external.theme.GC_Core_Subject` - so this issue would be tied to a user error.

The other thing - maybe it would be good to have a configuration so that `geonetwork.thesaurus` is configurable. Then, each geonetwork installation could determine what they want as a prefix.  There is already a catalogue name, so maybe, it would be better to use the `catalogue name` instead of hard coding it to `geonetwork.thesaurus`.  I wonder what happens in cases where metadata is harvested from another geonetwork instance which has similar thesaurus name but different content in their thesaurus - This could be a bug.

Reply:
> The French character issues seem to just be with excel not knowing the encoding standard. If imported into excel with UTF-8 it seems fine. Opening the raw text also looks fine.

The issue seems to be fixed in gn 4.  I tested on the following site https://apps.titellus.net/geonetwork and noticed that the file is no longer UTF-8 but ANSI instead.  So we should be able to expect this to be fixed in the next upgrade.